### ISSUE TITLE
SQL Generation: IS DISTINCT FROM

### ISSUE THREAD

**Issue by FrigoEU** [urweb/urweb#223]
Hey,

I've been running into some performance problems with generated SQL JOIN's where one of the sides is nullable. You fall into [this case](https://github.com/urweb/urweb/blob/master/src/monoize.sml#L2357) and since I'm using PostgreSQL, the generated SQL becomes IS NOT DISTINCT FROM or NOT (IS DISTINCT FROM). I'm however wondering why this is being used instead of the SQL used in eg SQLITE: a = b OR (a IS NULL AND b IS NULL)? 

As far as I know, this is the same in PostgreSQL, and is (probably always?) faster. I think it's also simpler, but I might be missing some subtle difference. 

In my production case, the difference is the query taking 1.5s or 100ms, so that's quite big.

I'd be happy to do the changes and make a PR myself, I just want to check it before I do any work on it.

Simon

**Comment by achlipala** [comment]
The code you see generated is an attempt to get, say, ML-style semantics for equality tests on optional values, as opposed to the perhaps-confusing SQL semantics with nulls.  I preferred `IS NOT DISTINCT FROM` where available to avoid duplicating subterms, which could lead to superlinear blowup in query size, in certain extreme situations.

I'm surprised that the subterm-duplicating version is so much faster!  Is it just a historical accident of which rules are in the Postgres query planner?

**Comment by FrigoEU** [comment]
Yup the difference is big, if I had to guess it's because US NOT DISTINCT
FROM is way less common then simple equality (ignoring the difference in
semantics) and as such less optimized.

I can see your reasoning in terms of limiting query size and I'd be all for
it, but in this case I would opt for performance, especially because the
query size will only very rarely be a problem I think.
